### PR TITLE
doc and test of security_group for change name use-case

### LINKED ISSUE #23708: Renaming Security Group, doesn't remove old SG's from resource first
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
Terraform v1.1.7
on darwin_amd64
+ provider registry.terraform.io/hashicorp/archive v2.2.0
+ provider registry.terraform.io/hashicorp/aws v4.5.0
+ provider registry.terraform.io/hashicorp/null v3.1.0
```

### Affected Resource(s)

* aws_security_group
* aws_instance

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

Original Config
```hcl
resource "aws_security_group" "public-http-https" {
  name_prefix = "lib"
  vpc_id = var.LibVPC
  tags = {
    Name = "http/https public"
    terraform = "Managed by Terraform"

  }
  ingress = [
    {
      cidr_blocks = [
        "0.0.0.0/0",
      ]
      description      = "Public HTTP Traffic"
      from_port        = 80
      to_port          = 80
      protocol         = "tcp"
      ipv6_cidr_blocks = []
      prefix_list_ids  = []
      security_groups  = []
      self             = false
    },
    {
      cidr_blocks = [
        "0.0.0.0/0",
      ]
      description      = "Public HTTPS Traffic"
      from_port        = 443
      to_port          = 443
      protocol         = "tcp"
      ipv6_cidr_blocks = []
      prefix_list_ids  = []
      security_groups  = []
      self             = false
    }
  ]

  egress {
    from_port        = 0
    to_port          = 0
    protocol         = "-1"
    cidr_blocks      = ["0.0.0.0/0"]
    ipv6_cidr_blocks = ["::/0"]
  }
}
```

```hcl
resource "aws_security_group" "public-http-https" {
  name = "libHttpHttpsPublic"
  vpc_id = var.LibVPC
  tags = {
    Name = "http/https public"
    terraform = "Managed by Terraform"

  }
  ingress = [
    {
      cidr_blocks = [
        "0.0.0.0/0",
      ]
      description      = "Public HTTP Traffic"
      from_port        = 80
      to_port          = 80
      protocol         = "tcp"
      ipv6_cidr_blocks = []
      prefix_list_ids  = []
      security_groups  = []
      self             = false
    },
    {
      cidr_blocks = [
        "0.0.0.0/0",
      ]
      description      = "Public HTTPS Traffic"
      from_port        = 443
      to_port          = 443
      protocol         = "tcp"
      ipv6_cidr_blocks = []
      prefix_list_ids  = []
      security_groups  = []
      self             = false
    }
  ]

  egress {
    from_port        = 0
    to_port          = 0
    protocol         = "-1"
    cidr_blocks      = ["0.0.0.0/0"]
    ipv6_cidr_blocks = ["::/0"]
  }
}
```

### Debug Output
n/a

### Expected Behavior

1. Create new security group
2. add security group to ec2 instance
3. remove old SG from ec2 instance
4. Delete old SG

### Actual Behavior

1. Attempts to delete old security group, without removing from instances

```
aws_security_group.public-http-https: Destroying... [id=sg-0836b287d3f086048]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 10s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 20s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 30s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 40s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 50s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 1m0s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 1m10s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 1m20s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 1m30s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 1m40s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 1m50s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 2m0s elapsed]
aws_security_group.public-http-https: Still destroying... [id=sg-0836b287d3f086048, 2m10s elapsed]
```

What happens is the security group attempts to get deleted before being removed from ec2 instances that are defined with the security group. If I go and remove the SG from the instance in the EC2 console, it continues as expected and eventually deletes it, creates the new one, and attaches it to the ec2 instance.

### Steps to Reproduce

1. `terraform apply`

### PR DESCRIPTION
**Pull request by bigmikes**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

### Changes
1. Added a new Acceptance Test to make sure that user can change `name` or `name-prefix` arguments to Security Group resource when they specify `create_before_destroy = true` property.
2. Extended Security Group documentation to provide this use-case as example. 

Closes #23708

Output from acceptance testing:

```
$ make testacc TESTS=TestAccEC2SecurityGroup_name_change PKG=ec2 ACCTEST_PARALLELISM=2  | tee test_output.txt
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 2 -run='TestAccEC2SecurityGroup_name_change'  -timeout 180m
=== RUN   TestAccEC2SecurityGroup_name_change
=== PAUSE TestAccEC2SecurityGroup_name_change
=== CONT  TestAccEC2SecurityGroup_name_change
--- PASS: TestAccEC2SecurityGroup_name_change (281.59s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	283.749s
```

### CODE REVIEW

File: website/docs/r/security_group.html.markdown [hunk 1]
```diff
@@ -96,6 +96,24 @@ resource "aws_vpc_endpoint" "my_endpoint" {
 
 You can also find a specific Prefix List using the `aws_prefix_list` data source.
 
+### Change of name or name-prefix value
+
+The `name` and `name-prefix` arguments force the creation of a new Security Group resource when they change value. In that case, Terraform first deletes the existing Security Group resource and then it creates a new one. If the existing Security Group is associated to a Network Interface resource, the deletion cannot complete. The reason is that Network Interface resources cannot be left with no Security Group attached and the new one is not yet available at that point.
+
+So, it is required to invert the default behavior of Terraform. That is, first the new Security Group resource must be created, then associated to possible Network Interface resources and finally the old Security Group can be detached and deleted. To force this behavior, you must set the [create_before_destroy](https://www.terraform.io/language/meta-arguments/lifecycle#create_before_destroy) property:
```
Comment (AdamTylerLynch) [inline website/docs/r/security_group.html.markdown:103]: Please remove the word "So,".

File: website/docs/r/security_group.html.markdown [hunk 2]
```diff
@@ -96,6 +96,24 @@ resource "aws_vpc_endpoint" "my_endpoint" {
 
 You can also find a specific Prefix List using the `aws_prefix_list` data source.
 
+### Change of name or name-prefix value
+
+The `name` and `name-prefix` arguments force the creation of a new Security Group resource when they change value. In that case, Terraform first deletes the existing Security Group resource and then it creates a new one. If the existing Security Group is associated to a Network Interface resource, the deletion cannot complete. The reason is that Network Interface resources cannot be left with no Security Group attached and the new one is not yet available at that point.
```
Comment (AdamTylerLynch) [inline website/docs/r/security_group.html.markdown:101]: If you can link to AWS public documentation to cite this, that would be a welcomed addition (but not required).

File: website/docs/r/security_group.html.markdown [hunk 3]
```diff
@@ -96,6 +96,24 @@ resource "aws_vpc_endpoint" "my_endpoint" {
...
+
+So, it is required to invert the default behavior of Terraform. That is, first the new Security Group resource must be created, then associated to possible Network Interface resources and finally the old Security Group can be detached and deleted. To force this behavior, you must set the [create_before_destroy](https://www.terraform.io/language/meta-arguments/lifecycle#create_before_destroy) property:
+
+```terraform
+resource "aws_security_group" "sg_with_changeable_name" {
+  name   = "changeable-name"
+  # ... other configuration ...
+
+  lifecycle {
+	  # Necessary if changing 'name' or 'name_prefix' properties.
```
Comment (AdamTylerLynch) [inline website/docs/r/security_group.html.markdown:111]: Tab again, please change to spaces.

**Review by AdamTylerLynch — CHANGES_REQUESTED** [review]
Small formatting changes requested. Thank you for providing clear documentation on how to get around this issue!

**Comment by bigmikes** [comment]
New PR: https://github.com/hashicorp/terraform-provider-aws/pull/24508

**Pull request closed without merging**